homeassistant/components/recorder/core.py

```diff
@@ -806,37 +803,14 @@ def _activate_and_set_db_ready(self) -> None:
                 for row in execute_stmt_lambda_element(session, get_migration_changes())
             }
 
-            migrator: BaseRunTimeMigration
-            for migrator_cls in (StatesContextIDMigration, EventsContextIDMigration):
+            for migrator_cls in (
+                StatesContextIDMigration,
+                EventsContextIDMigration,
+                EventTypeIDMigration,
+                EntityIDMigration,
+            ):
                 migrator = migrator_cls(session, schema_version, migration_changes)
-                if migrator.needs_migrate():
-                    self.queue_task(migrator.task())
-
-            migrator = EventTypeIDMigration(session, schema_version, migration_changes)
-            if migrator.needs_migrate():
-                self.queue_task(migrator.task())
-            else:
-                _LOGGER.debug("Activating event_types manager as all data is migrated")
-                self.event_type_manager.active = True
-
-            migrator = EntityIDMigration(session, schema_version, migration_changes)
-            if migrator.needs_migrate():
-                self.queue_task(migrator.task())
-            else:
-                _LOGGER.debug("Activating states_meta manager as all data is migrated")
-                self.states_meta_manager.active = True
-                with contextlib.suppress(SQLAlchemyError):
-                    # If ix_states_entity_id_last_updated_ts still exists
-                    # on the states table it means the entity id migration
-                    # finished by the EntityIDPostMigrationTask did not
-                    # because they restarted in the middle of it. We need
-                    # to pick back up where we left off.
-                    if get_index_by_name(
-                        session,
-                        TABLE_STATES,
-                        LEGACY_STATES_ENTITY_ID_LAST_UPDATED_INDEX,
-                    ):
-                        self.queue_task(EntityIDPostMigrationTask())
+                migrator.do_migrate(self)
 
             if self.schema_version > LEGACY_STATES_EVENT_ID_INDEX_SCHEMA_VERSION:
                 with contextlib.suppress(SQLAlchemyError):
```

Comment:
This should perhaps be handled by a `BaseRunTimeMigration` class instead?

Reply:
That would be nicer

Reply:
OK, I'll add that to a follow-up to not make this PR larger